### ISSUE TITLE
[No convergence] Convergence can not be reached when combined Encoder and Decoder

### ISSUE THREAD

**Issue by CaptainEven** [nianticlabs/monodepth2#338]
Hi, Thanks for your excellent work! I got a no-convergence problem when merge Encoder and Decoder into a single model like this (Stereo training of kitti):
```
class StereoNet(nn.Module):
    """
    PyTorch module for a resnet encoder
    """

    def __init__(self, num_layers,
                 pre_trained,
                 num_input_images=1,
                 num_ch_enc=[64, 64, 128, 256, 512],  # encode params
                 scales=range(4),
                 num_output_channels=1,
                 use_skips=True):  # decoder params
        """
        :param num_layers:
        :param pre_trained:
        :param num_input_images:
        """
        super(StereoNet, self).__init__()

        self.num_ch_enc = np.array(num_ch_enc)

        # ---------- encoder
        resnets = {
            18: models.resnet18,
            34: models.resnet34,
            50: models.resnet50,
            101: models.resnet101,
            152: models.resnet152
        }

        if num_layers not in resnets:
            raise ValueError("{} is not a valid number of resnet layers".format(num_layers))

        if num_input_images > 1:
            self.encoder = resnet_multiimage_input(num_layers, pre_trained, num_input_images)
        else:
            self.encoder = resnets[num_layers](pre_trained)
        print('Using models from {:s}.'.format('pre-trained' if pre_trained else 'stratch.'))

        if num_layers > 34:
            self.num_ch_enc[1:] *= 4

        ## ---------- decoder
        self.num_output_channels = num_output_channels
        self.use_skips = use_skips
        self.upsample_mode = 'nearest'
        self.scales = scales

        # self.num_ch_enc = num_ch_enc
        self.num_ch_dec = np.array([16, 32, 64, 128, 256])

        self.convs = OrderedDict()
        for i in range(4, -1, -1):
            # upconv_0
            num_ch_in = self.num_ch_enc[-1] if i == 4 else self.num_ch_dec[i + 1]
            num_ch_out = self.num_ch_dec[i]
            self.convs[("upconv", i, 0)] = ConvBlock(num_ch_in, num_ch_out)

            # upconv_1
            num_ch_in = self.num_ch_dec[i]
            if self.use_skips and i > 0:
                num_ch_in += self.num_ch_enc[i - 1]
            num_ch_out = self.num_ch_dec[i]
            self.convs[("upconv", i, 1)] = ConvBlock(num_ch_in, num_ch_out)

        for s in self.scales:
            self.convs[("dispconv", s)] = Conv3x3(self.num_ch_dec[s], self.num_output_channels)

        self.decoder = nn.ModuleList(list(self.convs.values()))
        self.sigmoid = nn.Sigmoid()

    def get_encoder_features(self, input_image):
        """
        :param input_image:
        :return:
        """
        ## ---------- encoder
        self.features = []

        x = (input_image - 0.45) / 0.225
        x = self.encoder.conv1(x)
        x = self.encoder.bn1(x)

        self.features.append(self.encoder.relu(x))  # /2
        self.features.append(self.encoder.layer1(self.encoder.maxpool(self.features[-1])))  # /4
        self.features.append(self.encoder.layer2(self.features[-1]))  # /8
        self.features.append(self.encoder.layer3(self.features[-1]))  # /16
        self.features.append(self.encoder.layer4(self.features[-1]))  # /32

        return self.features

    def get_decoder_depths(self, input_features):
        """
        :param input_features:
        :return:
        """
        # ---------- decoder
        self.outputs = {}

        # decoder
        x = input_features[-1]
        for i in range(4, -1, -1):
            x = self.convs[("upconv", i, 0)](x)
            x = [upsample(x)]

            if self.use_skips and i > 0:
                x += [input_features[i - 1]]

            x = torch.cat(x, 1)
            x = self.convs[("upconv", i, 1)](x)

            ## the last layer of each scale
            if i in self.scales:
                self.outputs[("disp", i)] = self.sigmoid(self.convs[("dispconv", i)](x))

        return self.outputs

    def forward(self, input_image):
        """
        :param input_image:
        :return:
        """
        ## ---------- encoder
        self.features = []

        x = (input_image - 0.45) / 0.225
        x = self.encoder.conv1(x)
        x = self.encoder.bn1(x)

        self.features.append(self.encoder.relu(x))
        self.features.append(self.encoder.layer1(self.encoder.maxpool(self.features[-1])))
        self.features.append(self.encoder.layer2(self.features[-1]))
        self.features.append(self.encoder.layer3(self.features[-1]))
        self.features.append(self.encoder.layer4(self.features[-1]))

        ## ---------- decoder
        self.outputs = {}

        x = self.features[-1]
        for i in range(4, -1, -1):
            x = self.convs[("upconv", i, 0)](x)
            x = [upsample(x)]

            if self.use_skips and i > 0:
                x += [self.features[i - 1]]

            x = torch.cat(x, 1)
            x = self.convs[("upconv", i, 1)](x)

            ## the last layer of each scale
            if i in self.scales:
                self.outputs[("disp", i)] = self.sigmoid(self.convs[("dispconv", i)](x))

        return self.outputs

        ## ----- Get features
        # self.get_encoder_features(input_image)

        # # ----- Get outputs
        # self.get_decoder_depths(self.features)

        # return self.outputs

```

**Comment by CaptainEven** [comment]
And the loss fluctuate around 0.2 like this:

![image](https://user-images.githubusercontent.com/9303747/116220405-3b268f00-a77f-11eb-9e19-0943159c7e20.png)

**Comment by CaptainEven** [comment]
This is how i define the Kitti dataset:
```
class MyKittiDataset(MonoDataset):
    def __init__(self, *args, **kwargs):
        """
        :param args:
        :param kwargs:
        """
        super(MyKittiDataset, self).__init__(*args, **kwargs)

        # NOTE: Make sure your intrinsics matrix is *normalized* by the original image size
        K = np.array([[718.335, 0.0, 609.5593, 0.0],
                      [0.0, 718.335, 172.8540, 0.0],
                      [0.0, 0.0, 1.0, 0.0],
                      [0.0, 0.0, 0.0, 1.0]], dtype=np.float32)
        self.K = np.eye(4, dtype=np.float32)
        self.K[0] = K[0] / self.img_width
        self.K[1] = K[1] / self.img_height

        # self.K[0, 2] = 0.5
        # self.K[1, 2] = 0.5

        self.full_res_shape = (self.img_width, self.img_height)
        self.side_map = {"2": 2, "3": 3, "l": 2, "r": 3}

        self.img_ext = '.jpg'  # Apollo use jpg image format

    def check_depth(self):
        """
        :return:
        """
        return False

    def get_image_path(self, folder, frame_idx, side):
        """
        :param folder:
        :param frame_idx:
        :param side:
        :return:
        """
        f_str = "{:010d}{}".format(frame_idx, self.img_ext)
        image_path = os.path.join(self.data_path,
                                  folder,
                                  "image_0{}/data/".format(self.side_map[side]),
                                  f_str)

        return image_path

    def get_color(self, folder, frame_idx, side, do_flip):
        """
        :param folder:
        :param frame_idx:
        :param side:
        :param do_flip:
        :return:
        """
        color = self.loader(self.get_image_path(folder, frame_idx, side))
        if do_flip:
            color = color.transpose(pil.FLIP_LEFT_RIGHT)

        return color
```

**Comment by mrharicot** [comment]
Hi,
Why do you want to merge both models into one module?
You can simply do so like this:
```python
class Monodepth2(nn.Module):
    def __init__(self, opt):
        super(Monodepth2, self).__init__()
        self.opt = opt
        self.encoder = networks.ResnetEncoder(self.opt.num_layers, self.opt.weights_init == "pretrained")
        self.decoder = networks.DepthDecoder(self.encoder.num_ch_enc, self.opt.scales)
        
    def forward(self, inputs):
        features = self.encoder(inputs["color_aug", 0, 0])
        outputs = self.decoder(features)
        return outputs
```

**Comment by CaptainEven** [comment]
Hi, because i want to deploy MonoDepthV2 to a embedding device, as a result the IO performance becomes a critical problem if intermediate data exists under the circumstance Encoder and Decoder separate.
Thanks for the advice and help! I will try it and give a reply as soon as possible.

**Comment by dexter2406** [comment]
> Hi, because i want to deploy MonoDepthV2 to a embedding device, as a result the IO performance becomes a critical problem if intermediate data exists under the circumstance Encoder and Decoder separate.
> Thanks for the advice and help! I will try it and give a reply as soon as possible.

It's just epoch_0, right? In my case, it took about 5 epochs to see a visible improvement. Also, the loss is pretty small,  so it's normal that it doesn't change much between epochs. But when the loss approaches ~1.0 you will see a fairly good result.

BTW, You can also check the summary and see if it's getting better qualitatively.

**Comment by CaptainEven** [comment]
@mrharicot @dexter2406 Thank you guys for the helpful advice! It's getting converging now, though very slowly. The problem may be caused by resetting the features and outputs at the time Encoder and Decoder is called. Further verification will be done, and i will give a feed back as soon as possible.

**Comment by mrharicot** [comment]
Great to hear it's converging! I will close this issue for now.
